File: v170/cgi-bin/webnotes/modules/SRM/QA Inspection Report.js
cur_frm.fields_dict['item_code'].get_query = function(doc) {
  if(doc.inspection_type == 'Incoming Stock Inspection'){
    return 'SELECT DISTINCT `tabPurchase Receipt Detail`.item_code FROM `tabPurchase Receipt Detail` WHERE `tabPurchase Receipt Detail`.parent = "' + doc.purchase_receipt_no + '" and `tabPurchase Receipt Detail`.qa_reported = "No" and `tabPurchase Receipt Detail`.item_code like "%s" ORDER BY `tabPurchase Receipt Detail`.item_code LIMIT 50';
  }
  else if(doc.inspection_type == 'Outgoing Stock Inspection'){
    return 'SELECT DISTINCT `tabDelivery Note Detail`.item_code FROM `tabDelivery Note Detail` WHERE `tabDelivery Note Detail`.parent = "' + doc.delivery_note_no + '" and `tabDelivery Note Detail`.item_code like "%s" ORDER BY `tabDelivery Note Detail`.item_code LIMIT 50';
  }
  else if(doc.inspection_type == 'InProcess Inspection'){
    return 'SELECT DISTINCT `tabItem`.name FROM `tabItem` WHERE `tabItem`.name like "%s" ORDER BY `tabItem`.name LIMIT 50';
  }
}

cur_frm.cscript.purchase_receipt_no = function(doc, cdt, cdn) {
  if (doc.purchase_receipt_no)
    get_server_fields('get_purchase_receipt_details','','',doc,cdt,cdn,1);
}

cur_frm.cscript.item_code = function(doc, cdt, cdn) {
  if (doc.item_code)
    get_server_fields('get_purchase_receipt_item_details','','',doc,cdt,cdn,1);
}

cur_frm.cscript.inspection_type = function(doc, cdt, cdn) {
  if(doc.inspection_type == 'Incoming Stock Inspection'){
    doc.delivery_note_no = '';
    doc.delivery_note_date = '';
    doc.delivery_note_detail = '';
    
    unhide_field('purchase_receipt_no');
    unhide_field('purchase_receipt_date');
    unhide_field('purchase_receipt_detail_no');
    hide_field('delivery_note_no');
    hide_field('delivery_note_date');
    hide_field('delivery_note_detail');
  }
  else if(doc.inspection_type == 'Outgoing Stock Inspection'){
    doc.purchase_receipt_no = '';
    doc.purchase_receipt_date = '';
    doc.purchase_receipt_detail_no = '';
    
    hide_field('purchase_receipt_no');
    hide_field('purchase_receipt_date');
    hide_field('purchase_receipt_detail_no');
    unhide_field('delivery_note_no');
    unhide_field('delivery_note_date');
    unhide_field('delivery_note_detail');
  }
  else if(doc.inspection_type == 'InProcess Inspection'){
    doc.purchase_receipt_no = '';
    doc.purchase_receipt_date = '';
    doc.purchase_receipt_detail_no = '';
    doc.delivery_note_no = '';
    doc.delivery_note_date = '';
    doc.delivery_note_detail = '';
    
    hide_field('purchase_receipt_no');
    hide_field('purchase_receipt_date');
    hide_field('purchase_receipt_detail_no');
    hide_field('delivery_note_no');
    hide_field('delivery_note_date');
    hide_field('delivery_note_detail');
  }
}

File: modules/SRM/QA Inspection Report/QA Inspection Report.py
class DocType:
  def __init__(self, doc, doclist=[]):
    self.doc = doc
    self.doclist = doclist

  def get_purchase_receipt_details(self):
    pr = sql("select transaction_date from `tabPurchase Receipt` where name = '%s'" % ( self.doc.purchase_receipt_no))
    ret = {
      'purchase_receipt_date' : pr and cstr(pr[0][0]) or '',
    }
    return cstr(ret)

  def get_purchase_receipt_item_details(self):
    pr_detail = sql("select name, sum(received_qty), sum(qty), sum(rejected_qty), description from `tabPurchase Receipt Detail` where item_code = '%s' and parent = '%s' group by item_code" % (self.doc.item_code, self.doc.purchase_receipt_no))
    ret = {
      'purchase_receipt_detail_no': pr_detail and pr_detail[0][0] or '',
      'description'  : pr_detail and pr_detail[0][4] or '',
      'receipt_qty'  : pr_detail and flt(pr_detail[0][1]) or 0,
      'checked_qty'  : pr_detail and flt(pr_detail[0][1]) or 0,
      'accepted_qty' : pr_detail and flt(pr_detail[0][2]) or 0,
      'rejected_qty' : pr_detail and flt(pr_detail[0][3]) or 0
    }
    return cstr(ret)

  def get_item_specification_details(self):
    self.doc.clear_table(self.doclist, 'qa_specification_details')
    specification = sql("select specification, value from `tabItem Specification Detail` where parent = '%s' order by idx" % (self.doc.item_code))
    for d in specification:
      child = addchild(self.doc, 'qa_specification_details', 'QA Specification Detail', 1, self.doclist)
      child.specification = d[0]
      child.value = d[1]
      child.status = 'Accepted'
        
  def on_submit(self):
    sql("update `tabPurchase Receipt Detail` set qa_reported = 'Yes' where name = '%s'" % (self.doc.purchase_receipt_detail_no))

  def on_cancel(self):
    sql("update `tabPurchase Receipt Detail` set qa_reported = 'No' where name = '%s'" % (self.doc.purchase_receipt_detail_no))